### PR TITLE
IGNITE-21419 Optimize listener collection update in AbstractEventProducer

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-21419

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java [hunk 1]
```diff
@@ -39,30 +41,24 @@ public abstract class AbstractEventProducer<T extends Event, P extends EventPara
...
         listenersByEvent.compute(evt, (evt0, listeners) -> {
-            List<EventListener<P>> newListeners;
-
-            if (listeners == null) {
-                newListeners = new ArrayList<>(1);
-            } else {
-                newListeners = new ArrayList<>(listeners.size() + 1);
-
-                newListeners.addAll(listeners);
+            if (nullOrEmpty(listeners)) {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java:44]: I suggest to never have empty lists in this map, so we can also dump the `isEmpty` check here

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java [hunk 2]
```diff
@@ -39,30 +41,24 @@ public abstract class AbstractEventProducer<T extends Event, P extends EventPara
...
-            return unmodifiableList(newListeners);
+            return addWithCopyList(listeners, (EventListener<P>) listener);
         });
     }
 
     @Override
     public void removeListener(T evt, EventListener<? extends P> listener) {
         listenersByEvent.computeIfPresent(evt, (evt0, listeners) -> {
-            var newListeners = new ArrayList<>(listeners);
+            if (nullOrEmpty(listeners)) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java:55]: Fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java [hunk 3]
```diff
@@ -39,30 +41,24 @@ public abstract class AbstractEventProducer<T extends Event, P extends EventPara
...
+                return null;
+            }
 
-            newListeners.remove(listener);
+            int indexOf = listeners.indexOf(listener);
 
-            return newListeners.isEmpty() ? null : unmodifiableList(newListeners);
+            return indexOf < 0 ? listeners : removeWithCopyList(listeners, indexOf);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java:61]: Fix it

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java (hunk 1 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java:44]: fix it

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 4]
```diff
@@ -592,4 +594,52 @@ public T2 next() {
...
+
+    /**
+     * Appends the specified element to the end of copy of list.
+     *
+     * @param <T> Type of elements of the list.
+     * @param list List.
+     * @param t Element to append.
+     * @return Immutable list with an appended element.
+     */
+    public static <T> List<T> addWithCopyList(List<? extends T> list, T t) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:606]: I think they might be useful, I'd prefer to keep them.

**Pull request closed without merging**